base/commands/migration/start_stages.go

```diff
@@ -0,0 +1,171 @@
+			Func:        st.startStage(ctx, ec),
+		},
+		{
+			ProgressMsg: "Migrating the cluster",
+			SuccessMsg:  "Migrated the cluster",
+			FailureMsg:  "Could not migrate the cluster",
+			Func:        st.migrateStage(ctx, ec),
+		},
+	}
+}
+
+func (st *StartStages) connectStage(ctx context.Context, ec plug.ExecContext) func(stage.Statuser) error {
+	return func(status stage.Statuser) error {
+		var err error
+		st.ci, err = ec.ClientInternal(ctx)
+		if err != nil {
+			return err
+		}
+		st.startQueue, err = st.ci.Client().GetQueue(ctx, StartQueueName)
+		if err != nil {
+			return err
+		}
+		st.statusMap, err = st.ci.Client().GetMap(ctx, MakeStatusMapName(st.migrationID))
+		if err != nil {
+			return err
+		}
+		st.updateTopic, err = st.ci.Client().GetTopic(ctx, MakeUpdateTopicName(st.migrationID))
+		if err != nil {
+			return err
+		}
+		st.updateMessageChan = make(chan UpdateMessage)
+		_, err = st.updateTopic.AddMessageListener(ctx, st.topicListener)
+		return err
+	}
+}
+
+func (st *StartStages) topicListener(event *hazelcast.MessagePublished) {
+	st.updateMessageChan <- event.Value.(UpdateMessage)
+}
+
+func (st *StartStages) startStage(ctx context.Context, ec plug.ExecContext) func(stage.Statuser) error {
+	return func(stage.Statuser) error {
+		cb, err := makeConfigBundle(st.configDir, st.migrationID)
+		if err != nil {
+			return err
+		}
+		if err = st.startQueue.Put(ctx, cb); err != nil {
+			return err
+		}
+		ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+		defer cancel()
+		if isTerminal, err := st.handleUpdateMessage(ctx, ec, <-st.updateMessageChan); isTerminal {
+			return err
+		}
+		return nil
+	}
+}
+
+func makeConfigBundle(configDir, migrationID string) (serialization.JSON, error) {
+	var cb ConfigBundle
+	cb.MigrationID = migrationID
+	if err := cb.Walk(configDir); err != nil {
+		return nil, err
+	}
+	b, err := json.Marshal(cb)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func (st *StartStages) migrateStage(ctx context.Context, ec plug.ExecContext) func(statuser stage.Statuser) error {
+	return func(stage.Statuser) error {
+		defer st.updateTopic.RemoveListener(ctx, st.topicListenerID)
+		for {
+			select {
+			case msg := <-st.updateMessageChan:
+				if isTerminal, err := st.handleUpdateMessage(ctx, ec, msg); isTerminal {
+					return err
+				}
+			case <-ctx.Done():
+				if err := ctx.Err(); err != nil {
+					if errors.Is(err, context.DeadlineExceeded) {
+						return timeoutErr
+					}
+					return fmt.Errorf("migration failed: %w", err)
+				}
+			}
+		}
+	}
+}
+
+func (st *StartStages) handleUpdateMessage(ctx context.Context, ec plug.ExecContext, msg UpdateMessage) (bool, error) {
+	ec.PrintlnUnnecessary(msg.Message)
+	if slices.Contains([]Status{StatusComplete, StatusFailed, StatusCanceled}, msg.Status) {
+		ms, err := readMigrationStatus(ctx, st.statusMap)
+		if err != nil {
+			return true, fmt.Errorf("reading status: %w", err)
+		}
+		ec.PrintlnUnnecessary(ms.Report)
```

Comment:
The report should also be printed to a file.

Reply:
As I said in the standup, let's do this in this PR as well.

Reply:
https://github.com/hazelcast/hazelcast-commandline-client/pull/367/commits/999a3bfc98a13046d24672251010b903dbff6604